nearcore/src/state_sync.rs

```diff
@@ -414,7 +429,116 @@ fn check_new_epoch(
             // Still in the latest dumped epoch. Do nothing.
             Ok(None)
         } else {
-            start_dumping(head.epoch_id, sync_hash, shard_id, &chain, runtime)
+            start_dumping(head.epoch_id, sync_hash, shard_id, &chain, runtime, account_id)
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::state_sync::spawn_state_sync_dump;
+    use near_chain::{ChainGenesis, Provenance};
+    use near_chain_configs::{DumpConfig, ExternalStorageLocation};
+    use near_client::test_utils::TestEnv;
+    use near_network::test_utils::wait_or_timeout;
+    use near_o11y::testonly::init_test_logger;
+    use near_primitives::hash::CryptoHash;
+    use near_primitives::state_part::PartId;
+    use near_primitives::types::BlockHeight;
+    use std::ops::ControlFlow;
+    use std::str::FromStr;
+    use std::time::Duration;
+
+    #[test]
+    /// Produce several blocks, wait for the state dump thread to notice and
+    /// write files to a temp dir.
+    fn test_state_dump() {
+        init_test_logger();
+
+        let mut chain_genesis = ChainGenesis::test();
+        chain_genesis.epoch_length = 5;
+        let mut env = TestEnv::builder(chain_genesis.clone()).build();
+        let chain = &env.clients[0].chain;
+        let runtime = chain.runtime_adapter();
+        let mut config = env.clients[0].config.clone();
+        let root_dir = tempfile::Builder::new().prefix("state_dump").tempdir().unwrap();
+        config.state_sync_config_dump = Some(DumpConfig {
+            location: ExternalStorageLocation::Filesystem {
+                root_dir: root_dir.path().to_path_buf(),
+            },
+            restart_dump_for_shards: None,
+            iteration_delay: Some(Duration::from_millis(250)),
+        });
+
+        const MAX_HEIGHT: BlockHeight = 15;
+
+        near_actix_test_utils::run_actix(async move {
+            let _state_sync_dump_handle = spawn_state_sync_dump(
+                &config,
+                chain_genesis,
+                runtime.clone(),
+                Some("test0".parse().unwrap()),
+            )
+            .unwrap();
+            let mut last_block = None;
+            for i in 1..=MAX_HEIGHT {
+                let block = env.clients[0].produce_block(i as u64).unwrap().unwrap();
+                last_block = Some(block.clone());
+                env.process_block(0, block, Provenance::PRODUCED);
+            }
+            let epoch_id = runtime.get_epoch_id(last_block.clone().unwrap().hash()).unwrap();
+            let epoch_info = runtime.get_epoch_info(&epoch_id).unwrap();
+            let epoch_height = epoch_info.epoch_height();
+
+            let num_shards = runtime.num_shards(&epoch_id).unwrap();
+            assert_eq!(num_shards, 1);
+            assert_eq!(epoch_height, 10);
```

Comment:
this epoch height check doesn't seem like a good thing to test for in my opinion. Essentially we're testing that `KeyValueruntime::get_epoch_info()` indeed hardcodes 10 and not 11 or 12 or 3? Same with `num_shards`. Why not just build the filepaths below based on what the runtime returns?

Reply:
Done